docs/rules/expiring-todo-comments.md

```diff
@@ -0,0 +1,286 @@
+to the future where some feature will be widely available and the could drop support for
+older versions.
+
+Imagine you want to use async/await but for now you support node versions that don't have it?
+Maybe npm will have a feature that could be useful later but for now you have to keep support
+for older version users?
+
+```js
+// TODO [engine:node@>=8]: We can use async/await now.
+// FIXME [engine:node@>=20.0.0] Hey, node can use import/export now, we should refactor.
+// XXX @lubien [engine:npm@>4.0]: Now npm have [feature] we should use it.
+```
+
+Argument versions should be [semver](https://semver.org/) compatible such as: `1.2.45`, `5.3`, `1`.
+
+### Dependency Presence
+
+As a developer using NodeJS you should be know to package.json
+[dependencies](https://docs.npmjs.com/files/package.json#dependencies) and
+[devDependencies](https://docs.npmjs.com/files/package.json#devdependencies).
+This condition works over the presence or absence of dependencies in both fields.
+
+You may use `+` to trigger when the dependency is added and `-` when the dependency is
+removed.
+
+```js
+// TODO [-vue-function-api]: When we remove `vue-function-api` we should refactor this.
+// FIXME [+read-pkg] If we use this package we don't need to use this function below.
+// XXX @lubien [+react, -jquery]: We can use react for this widget instead of JQuery.
+```
+
+This condition is really useful when the code should be looked upon
+changes in dependencies that are known to happen such as `vue-function-api` that will
+eventually be part of `vue` itself.
+
+### Dependency Version
+
+Another way to make conditions over dependencies is to look upon versions. This is
+useful when developers know that further versions of dependencies make changes that
+impact on their code.
+
+```js
+// TODO [vue@>=3]: Refactor to function API when it's stable.
+// FIXME [cerebro@>0.10.0] This is a quickfix until cerebro fix this.
+// XXX [popura@>=2.0.0] This API is deprecated so we should not use it by then.
+```
+
+Argument versions should be [semver](https://semver.org/) compatible such as: `1.2.45`, `5.3`, `1`.
+
+### Combinations
+
+Any combination of rules is possible as long as you separate by commas. Each
+condition **trigger an individual report.**
+
+```js
+// TODO [+react, -jquery]: We can use react for this widget instead of JQuery.
+// TODO [2019-07-15, +react]: Refactor this if we install React or if we reach that date.
+// TODO [-vue-function-api, vue@>=3]: Now we should use Vue native function API.
+```
+
+## Disallow Warning Comments (no-warning-comments)
+
+This rule implements [`eslint/no-warning-comments`](https://eslint.org/docs/rules/no-warning-comments).
+
+The sole difference is that first we check for **valid conditions**
+to apply this rule. If no valid condition is met we fallback to
+`eslint/no-warning-comments` and you'll see something like
+`Unexpected 'todo' comment.`
```

Comment:
I'm wondering if we should add a setting to disable this. Imagine you would want condition TODO comment to error, but normal TODO comments to just warn. Unfortunately, there's no way for rules to force a report to be a warning. So the best we can do is to allow disabling it, so users can use the `eslint/no-warning-comments` rule in addition with a warning level.

Reply:
How about: `allowWarningComments`? Defaults to true?

Reply:
Done